src/dbnode/storage/series/buffer.go

```diff
@@ -473,119 +440,346 @@ func (b *dbBuffer) FetchBlocksMetadata(
+		// Check if we have the same pointer in cache.
+		if buckets == newBuckets {
+			replaceIdx = i
+		}
+	}
 
-	bopts := b.opts.DatabaseBlockOptions()
-	encoder := bopts.EncoderPool().Get()
-	encoder.Reset(start, bopts.DatabaseBlockAllocSize())
+	for i := replaceIdx; i > 0; i-- {
+		b.bucketVersionsCache[i] = b.bucketVersionsCache[i-1]
+	}
+
+	b.bucketVersionsCache[0] = newBuckets
+}
+
+func (b *dbBuffer) removeBucketVersionsInCache(oldBuckets *BufferBucketVersions) {
+	nilIdx := -1
+	for i, buckets := range b.bucketVersionsCache {
+		if buckets == oldBuckets {
+			nilIdx = i
+		}
+	}
+	if nilIdx == -1 {
+		return
+	}
+
+	for i := nilIdx; i < bucketsCacheSize-1; i++ {
+		b.bucketVersionsCache[i] = b.bucketVersionsCache[i+1]
+	}
+
+	b.bucketVersionsCache[bucketsCacheSize-1] = nil
+}
+
+func (b *dbBuffer) removeBucketVersionsAt(blockStart time.Time) {
+	buckets, exists := b.bucketVersionsAt(blockStart)
+	if !exists {
+		return
+	}
+	delete(b.bucketsMap, xtime.ToUnixNano(blockStart))
+	b.removeBucketVersionsInCache(buckets)
+	b.inOrderBlockStartsRemove(blockStart)
+	// nil out pointers.
+	buckets.resetTo(timeZero, nil, nil)
+	b.bucketVersionsPool.Put(buckets)
+}
+
+func (b *dbBuffer) inOrderBlockStartsAdd(newTime time.Time) {
+	starts := b.inOrderBlockStarts
+	idx := len(starts)
+	// There shouldn't be that many starts here, so just linear search through.
+	for i, t := range starts {
+		if t.After(newTime) {
+			idx = i
+			break
+		}
+	}
+	// Insert new time without allocating new slice.
+	b.inOrderBlockStarts = append(starts, timeZero)
+	// Update to new slice
+	starts = b.inOrderBlockStarts
+	copy(starts[idx+1:], starts[idx:])
+	starts[idx] = newTime
+}
 
+func (b *dbBuffer) inOrderBlockStartsRemove(removeTime time.Time) {
+	starts := b.inOrderBlockStarts
+	// There shouldn't be that many starts here, so just linear search through.
+	for i, t := range starts {
+		if t.Equal(removeTime) {
+			b.inOrderBlockStarts = append(starts[:i], starts[i+1:]...)
+			return
+		}
+	}
+}
+
+// BufferBucketVersions is a container for different versions (from
+// different flushes to disk) of buffer buckets in the database.
+type BufferBucketVersions struct {
+	buckets           []*BufferBucket
+	start             time.Time
+	opts              Options
+	lastReadUnixNanos int64
+	bucketPool        *BufferBucketPool
+}
+
+func (b *BufferBucketVersions) resetTo(
+	start time.Time,
+	opts Options,
+	bucketPool *BufferBucketPool,
+) {
+	// nil all elements so that they get GC'd.
+	for i := range b.buckets {
+		b.buckets[i] = nil
+	}
+	b.buckets = b.buckets[:0]
 	b.start = start
-	b.encoders = append(b.encoders, inOrderEncoder{
-		encoder: encoder,
-	})
-	b.bootstrapped = nil
+	b.opts = opts
 	atomic.StoreInt64(&b.lastReadUnixNanos, 0)
-	b.drained = false
+	b.bucketPool = bucketPool
 }
 
-func (b *dbBufferBucket) finalize() {
-	b.resetEncoders()
-	b.resetBootstrapped()
+func (b *BufferBucketVersions) streams(ctx context.Context, wType WriteType) []xio.BlockReader {
```

Comment:
This method is an if/else on the `WriteType` but I see a few places you call this twice (to get a result for each WriteType) and then concat them together. 

Would it perhaps be better to create an options struct so you can just call it once for those cases? i.e.

```
type streamsOptions struct {
  filterWriteType bool
  writeType WriteType
}

func (b *BufferBucketVersions) streams(ctx context.Context, opts streamsOptions) []xio.BlockReader {
	var res []xio.BlockReader
	for _, bucket := range b.buckets {
		if !opts.filterWriteType || bucket.wType == opts.writeType {
			res = append(res, bucket.streams(ctx)...)
		}
	}

 	return res
}
```

Reply:
Good idea, I like it.